migration/bucket_migrator.go

```diff
@@ -0,0 +1,106 @@
+package migration
+
+import (
+	"bytes"
+
+	"github.com/NethermindEth/juno/db"
+)
+
+var _ Migration = (*BucketMigrator)(nil)
+
+type (
+	BucketMigratorDoFunc    func(t db.Transaction, b1, b2 []byte) error
+	BucketMigratorKeyFilter func([]byte) (bool, error)
+)
+
+type BucketMigrator struct {
+	target db.Bucket
+
+	// number of entries to update before returning migration.ErrCallWithNewTransaction
+	batchSize uint
+	before    func()
+	keyFilter BucketMigratorKeyFilter
+	do        BucketMigratorDoFunc
+
+	// key to seek to when starting the migration
+	startFrom []byte
+}
+
+func NewBucketMigrator(target db.Bucket, do BucketMigratorDoFunc) *BucketMigrator {
+	return &BucketMigrator{
+		target:    target,
+		startFrom: target.Key(),
+		batchSize: 1_000_000,
+
+		before:    func() {},
+		keyFilter: func(b []byte) (bool, error) { return true, nil },
+		do:        do,
+	}
+}
+
+func NewBucketMover(source, destination db.Bucket) *BucketMigrator {
+	return NewBucketMigrator(source, func(txn db.Transaction, key, value []byte) error {
+		err := txn.Delete(key)
+		if err != nil {
+			return err
+		}
+
+		key[0] = byte(destination)
+		return txn.Set(key, value)
+	})
+}
+
+func (m *BucketMigrator) WithBatchSize(batchSize uint) *BucketMigrator {
+	m.batchSize = batchSize
+	return m
+}
+
+func (m *BucketMigrator) WithBefore(before func()) *BucketMigrator {
+	m.before = before
+	return m
+}
+
+func (m *BucketMigrator) WithKeyFilter(keyFilter func([]byte) (bool, error)) *BucketMigrator {
+	m.keyFilter = keyFilter
+	return m
+}
+
+func (m *BucketMigrator) Before() {
+	m.before()
+}
+
+func (m *BucketMigrator) Migrate(txn db.Transaction) error {
+	remainingInBatch := m.batchSize
+	iterator, err := txn.NewIterator()
+	if err != nil {
+		return err
+	}
+
+	for iterator.Seek(m.startFrom); iterator.Valid(); iterator.Next() {
+		key := iterator.Key()
```

Comment:
We are doing some interesting things with `key` later on. Should we create a defensive copy to be safe? Same with `value`.

I think we should at least document somewhere that `key` and `value` should not be mutated.

Reply:
`key` and `value` are copies, so the caller owns these slices.